phoenix-core-client/src/main/java/org/apache/phoenix/util/UpgradeUtil.java

```diff
@@ -1494,76 +1498,79 @@ public static void moveHBaseLevelTTLToSYSCAT(PhoenixConnection oldMetaConnection
                             rowKeys.add(tmpKey);
-                            String tableName = SchemaUtil.getTableName(rr.getValue(
-                                            DEFAULT_COLUMN_FAMILY_BYTES, TABLE_SCHEM_BYTES),
-                                    rr.getValue(DEFAULT_COLUMN_FAMILY_BYTES, TABLE_NAME_BYTES));
-                            if (tableName == null || Arrays.equals(rr.getValue(DEFAULT_COLUMN_FAMILY_BYTES,
-                                    TABLE_SCHEM_BYTES), SYSTEM_SCHEMA_NAME_BYTES)) {
+                            byte[][] rowKeyMetaData = new byte[3][];
+                            getVarChars(tmpKey, 3, rowKeyMetaData);
+                            byte[] schemaName = rowKeyMetaData[PhoenixDatabaseMetaData.SCHEMA_NAME_INDEX];
+                            byte[] tableName = rowKeyMetaData[PhoenixDatabaseMetaData.TABLE_NAME_INDEX];
+
+                            String fullTableName = SchemaUtil.getTableName(schemaName, tableName);
+                            if (SchemaUtil.isSystemTable(SchemaUtil.getTableNameAsBytes(schemaName, tableName))) {
                                 //We do not support system table ttl through phoenix ttl, and it will be moved to a
                                 //constant value in future commit.
                                 continue;
                             }
                             TableDescriptor tableDesc = admin.getDescriptor(SchemaUtil.getPhysicalTableName(
-                                    tableName, readOnlyProps));
+                                    fullTableName, readOnlyProps));
                             int ttl = tableDesc.getColumnFamily(DEFAULT_COLUMN_FAMILY_BYTES).
                                     getTimeToLive();
-                            if (ttl != ColumnFamilyDescriptorBuilder.DEFAULT_TTL) {
-                                //As we have ttl defined fot this table create a Put to set TTL.
-                                long rowTS = rr.rawCells()[0].getTimestamp();
+                            // As we have ttl defined for this table create a Put to set TTL with
+                            // backward compatibility in mind.
+                            long rowTS = EnvironmentEdgeManager.currentTimeMillis();
+
+                            if (ttl == HConstants.FOREVER || ttl == TTL_NOT_DEFINED) {
+                                // Set the TTL column to null
+                                Delete deleteColumn = new Delete(tmpKey);
+                                KeyValue kv = GenericKeyValueBuilder.INSTANCE.buildDeleteColumns(
+                                        new ImmutableBytesWritable(tmpKey),
+                                        new ImmutableBytesWritable(PhoenixDatabaseMetaData.TABLE_FAMILY_BYTES),
+                                        new ImmutableBytesWritable(TTL_BYTES), rowTS);
+                                deleteColumn.add(kv);
+                                mutations.add(deleteColumn);
```

Comment:
Is this necessary ? For all the tables which don't have TTL defined we will be doing a  Delete mutation. This will just increase the time.

Reply:
The time will anyway be spent scanning SYSTEM.CATALOG to find tables with TTL. This Delete Mutation is just to conform to SYSCAT standards of inserting a DeleteColumn to indicate a null, It does for every other column/attribute that is null when inserting rows.

Reply:
Oh, you are saying that it is null to begin with - so no point in adding a null again? Let me verify and will remove it if that is the case.